### PR TITLE
jetbrains.jdk: openjdk11 (11.0.13-b1751.25) → openjdk17 (17.0.3-b469.37)

### PR DESCRIPTION
**Pull request by fabianhjr**
###### Description of changes

Two bundled updates, one of runtime (JBR) and IntelliJ IDEA Community Edition

I had been using the JBR 17 update for a while in previous versions of Idea Comumnity (only nuance were some warnings), with upstream moving the default forward this might be a good time to move nixpkgs default runtime forward too.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by AnatolyPopov** [comment]
Exactly the thing I was looking into today also! Will try to take a look soonish.

**Comment by fabianhjr** [comment]
Just as a note, other JetBrains packages should be tested too since the runtime update affect those.

**Comment by AnatolyPopov** [comment]
I was actually thinking about not dropping JBR11 but keeping both for sometime but will see

**Comment by drupol** [comment]
Are you going to update all the other IDEs ? Thanks.

**Comment by fabianhjr** [comment]
> I was actually thinking about not dropping JBR11 but keeping both for sometime but will see

AFAIK all previous versions of editors (at least 2022.1) work great with JRE 17, there might have been some kinks/rough edges but JetBrains decided to move the default forward so no need to keep the previous versions.

**Comment by fabianhjr** [comment]
> Are you going to update all the other Jetbrains IDEs ? Thanks.

Not really, I don't use any other editor and this PR hasn't gotten reviews/merged as it is.

**Comment by fabianhjr** [comment]
> @fabianhjr I did take a look into these things and for some reason I still managed to build JBR17 locally. It depends on openjdk17 that I believe is bootstrapped via openjdk17-bootstrap that is defined as
> 
> https://github.com/NixOS/nixpkgs/blob/511e56d76cf528afcbff2e5c7930a791e5484eee/pkgs/top-level/java-packages.nix#L126
> 
> So when I'm trying to build JBR17 I'm getting an error with wrong bootstrap version complaining about openjdk16 and asking for JDK 11.
> 
> [..]
>
> Or can I you say what am I doing wrong?

Seems like a darwin/macOS issue due to a typo, I believe it should be using version 17 to bootstrap.

https://github.com/NixOS/nixpkgs/commit/a3401f6e3356cbf76050ed276f7b26d87766dc26

**Comment by fabianhjr** [comment]
Ah, nvm, seems like there is something going on with JDK 17 and darwin/macOS.

https://github.com/NixOS/nixpkgs/pull/143836#issuecomment-978045999

I have no access to that platform, could you try to change that bootstrap version to 17 and check if it works? @AnatolyPopov

**Comment by fabianhjr** [comment]
Seems like adopt-openjdk 17 has been stalled on darwin/macOS since january?

https://github.com/NixOS/nixpkgs/pull/165354

Could you try to cherry-pick https://github.com/NixOS/nixpkgs/pull/165354/commits/af9e23d9c7edf2dbfa9cff7fb3e50ff6534781bb for testing tho adopt-openjdk-17 for bootstrapping?

**Comment by AnatolyPopov** [comment]
@fabianhjr Thanks for digging into this but first of all I also do not have MacOS/Darwin, I'm on NixOS.
Second - I would suggest waiting for adoptopenjdk17 to be merged, there are several related PRs in progress.
Third - when I tried to build I tried with the JCEF version of JBR and there seems to be some issues https://github.com/NixOS/nixpkgs/issues/160468

I will try to run the build with JBR17 without JCEF a little later and we'll see how it will go.

**Comment by drupol** [comment]
I would also skip PHPStorm 2022.2 and use directly 2022.2.1 when it's out, see https://blog.jetbrains.com/phpstorm/2022/08/phpstorm-2022-2-1-preview/

**Comment by AnatolyPopov** [comment]
@drupol I believe this PR will not include any updates except Intellij unfortunately. 
Also, @fabianhjr since there are some complications with JBR17 could we maybe create separate PR for regular version upgrade and keep this one focused on JBR? Lates Intellij anyways seems to be working fine with JBR11

**Comment by fabianhjr** [comment]
Done so, cherry-picked the "apdoptopenjdk: add 17.0.2" commit and change of bootstrap

File: pkgs/top-level/all-packages.nix
Comment (Mindavi) [inline pkgs/top-level/all-packages.nix:12569]: What does the 'inherit lib' do here?

**Review by Mindavi — APPROVED** [review]
Diff LGTM, one question

**Comment by Mindavi** [comment]
There's a hash mismatch on Darwin too, see ofborg.

File: pkgs/development/compilers/jetbrains-jdk/default.nix
Comment (fabianhjr) [inline pkgs/development/compilers/jetbrains-jdk/default.nix:16]: > There's a hash mismatch on Darwin too, see ofborg.

Hi, confused about this @Mindavi, it is from a git fetch with a specific revision. :O

A hash mismatch shouldn't be possible to begin with (specially since the linux x86_64 / aarch64 jobs succeeded.

File: pkgs/top-level/all-packages.nix
Comment (fabianhjr) [reply pkgs/top-level/all-packages.nix:12569]: This is cherry-picked from: https://github.com/NixOS/nixpkgs/pull/165354

afaik it is a common pattern on adoptopenjdk (check lines above/bellow following the same pattern) but I am not familiarized with why it is done that way.

File: pkgs/development/compilers/jetbrains-jdk/default.nix
Comment (Mindavi) [reply pkgs/development/compilers/jetbrains-jdk/default.nix:16]: It might be due to case (in)sensitive filesystems mismatching between Darwin and linux 😞

Comment (fabianhjr) [reply pkgs/development/compilers/jetbrains-jdk/default.nix:16]: :/

Is there a tool to generate the hash from linux?

Comment (Mindavi) [reply pkgs/development/compilers/jetbrains-jdk/default.nix:16]: Ofborg already did? :)

File: pkgs/top-level/all-packages.nix
Comment (Mindavi) [reply pkgs/top-level/all-packages.nix:12569]: Yeah, I saw that but I couldn't explain it, so I was wondering if you could :). What happens if it's removed?

File: pkgs/development/compilers/jetbrains-jdk/default.nix
Comment (fabianhjr) [reply pkgs/development/compilers/jetbrains-jdk/default.nix:16]: Ah right, so now only adding that hash and a switch between those two possible hashes. Will add a commit tomorrow morning addressing this. 

Thanks for the help. :3

Comment (uri-canva) [reply pkgs/development/compilers/jetbrains-jdk/default.nix:16]: Hash mismatch due to case insensitive + case preserving filesystems like darwin's and windows' would only happen if there were multiple files only differing in case, like a `Config` file and a `config` file. Might be worth double checking this isn't an issue. If you have access to a macOS machine you can do that by creating two volumes, a case insensitive one and a case sensitive one, and then diffing the extracted github tarball. On linux you can find ways of doing that on SO, like https://stackoverflow.com/questions/2109056/how-to-find-duplicate-files-with-same-name-but-in-different-case-that-exist-in-s for example.

Comment (thefloweringash) [reply pkgs/development/compilers/jetbrains-jdk/default.nix:16]: Looks like there are two identical files called `DrawTest.java` at nearly identical paths.

```
❯ curl -fsSL https://github.com/JetBrains/JetBrainsRuntime/archive/jb17.0.3-b469.32.tar.gz \
  | tar tz \
  | sort -f | uniq -D -i
JetBrainsRuntime-jb17.0.3-b469.32/test/jdk/jb/java/awt/font/
JetBrainsRuntime-jb17.0.3-b469.32/test/jdk/jb/java/awt/Font/
JetBrainsRuntime-jb17.0.3-b469.32/test/jdk/jb/java/awt/font/TextLayout/
JetBrainsRuntime-jb17.0.3-b469.32/test/jdk/jb/java/awt/Font/TextLayout/
JetBrainsRuntime-jb17.0.3-b469.32/test/jdk/jb/java/awt/font/TextLayout/DrawTest.java
JetBrainsRuntime-jb17.0.3-b469.32/test/jdk/jb/java/awt/Font/TextLayout/DrawTest.java
```

Comment (fabianhjr) [reply pkgs/development/compilers/jetbrains-jdk/default.nix:16]: `diff ./test/jdk/jb/java/awt/font/TextLayout/DrawTest.java test/jdk/jb/java/awt/Font/TextLayout/DrawTest.java` is empty, then an easier approach would be to delete one of the duplicates?

Comment (fabianhjr) [reply pkgs/development/compilers/jetbrains-jdk/default.nix]: Opened an issue upstream, going to go with the `hash = if macos then X else Y` approach.

**Review by drupol — APPROVED** [review]

File: pkgs/development/compilers/jetbrains-jdk/default.nix
Comment (uri-canva) [inline pkgs/development/compilers/jetbrains-jdk/default.nix:18]: Do we know if this is deterministic? I can't find any reference to whether git handles case conflicts in a deterministic way or not.

**Review by uri-canva — CHANGES_REQUESTED** [review]
The build on ofborg is failing, and I get the same failure locally:

```
error: builder for '/nix/store/svw1izkb2xmr18flk613cpdcsg8f0xhk-jetbrains-jdk-17.0.3-b469.37.drv' failed with exit code 1;
       last 10 log lines:
       > unpackPhase completed in 1 minutes 2 seconds
       > patching sources
       > configuring
       > configure flags: --prefix=/nix/store/5llzkmsmllzycs7a5bv9iwpxamszvxin-jetbrains-jdk-17.0.3-b469.37
       > Runnable configure script is not present
       > Generating runnable configure script at /private/tmp/nix-build-jetbrains-jdk-17.0.3-b469.37.drv-0/source/build/.configure-support/generated-configure.sh
       >
       > Autoconf is not found on the PATH (/nix/store/v4fybh0lri7qdq9bplq2hdp24a7lygd1-unzip-6.0/bin:/nix/store/h7066bap24anl14mmc3539xqfhhcsfyf-clang-wrapper-11.1.0/bin:/nix/store/lpysf2qn09j9nyd4l7s3xjww6qhmqa1i-clang-11.1.0/bin:/nix/store/3mfkgajns47hfv0diihzi2scwl4hm2fl-coreutils-9.1/bin:/nix/store/gw88v0jnfykhm3vgcy0c5q80ljip9i6x-cctools-binutils-darwin-wrapper-949.0.1/bin:/nix/store/nq7dispfy6jj8m56963b5w990djs431y-cctools-binutils-darwin-949.0.1/bin:/nix/store/3mfkgajns47hfv0diihzi2scwl4hm2fl-coreutils-9.1/bin:/nix/store/8d0wgshdr8qa81ylani8kyld6ap317cl-findutils-4.9.0/bin:/nix/store/4d236ml1h032wr2vglnldbjkapcrhw2d-diffutils-3.8/bin:/nix/store/vb661ycg7yvg7mvfy27z78j2dv29760w-gnused-4.8/bin:/nix/store/jdvpk1v5a1bbzfjm510b97p0bh7vnc5q-gnugrep-3.7/bin:/nix/store/1rkr80jzyilhw7mcasqfs9i8hcz3wpc0-gawk-5.1.1/bin:/nix/store/bi4ffmv3hxzpdy4jqcm15n36plmv6k7g-gnutar-1.34/bin:/nix/store/rjk7n53h36yzk55vy218nw32y3h51rkz-gzip-1.12/bin:/nix/store/mcj3cypgv0pramf11a3pwazlyg1iyfg8-bzip2-1.0.8-bin/bin:/nix/store/yflpngzrwq9imh90wy6p7r8jysg6x5w4-gnumake-4.3/bin:/nix/store/db7n4xkfng8y4c3klp6bjppaws3rphq6-bash-5.1-p16/bin:/nix/store/lmkwb03xcz7z4np5kanq56svm8k95kz2-patch-2.7.6/bin:/nix/store/dvvvwv9m34sylq63nnx0d4rjsbps773v-xz-5.2.5-bin/bin:/nix/store/q7nqswqxq8fjp2qd9yi9jw0qnw2vpc3g-file-5.42/bin), and AUTOCONF is not set.
       > You need autoconf to be able to generate a runnable configure script.
       > Error: Cannot find autoconf
       For full logs, run 'nix log /nix/store/svw1izkb2xmr18flk613cpdcsg8f0xhk-jetbrains-jdk-17.0.3-b469.37.drv'.
```

File: pkgs/development/compilers/jetbrains-jdk/default.nix
Comment (fabianhjr) [reply pkgs/development/compilers/jetbrains-jdk/default.nix:18]: I don't have access to a macOS device, could you help with that?

Comment (fabianhjr) [reply pkgs/development/compilers/jetbrains-jdk/default.nix:18]: Also, the hash of the two files that clash is the same. (It doesn't matter which of the two copies git chooses, it will hash the same)

**Comment by drupol** [comment]
What's the status of this?

**Comment by fabianhjr** [comment]
Not merged

**Comment by fabianhjr** [comment]
I cannot help on the macOS front, if a @NixOS/darwin-maintainers could help it would be appreciated.

Been using this for a while on NixOS though.

**Comment by fabianhjr** [comment]
Rebased due to #165354 merge

**Comment by drupol** [comment]
Current Jetbrains tools are broken when it comes to the use SSH remote connection, this is a huge problem for me and I guess for some other people as well.

It would be nice to move on with this. WDYT ?

**Comment by veprbl** [comment]
Seems like fcfdb620c89872c13320ffce3263bdc349a762eb fixes the fetching issue. But the build of openjfx fails on x86_64-linux.

**Comment by fabianhjr** [comment]
> Current Jetbrains tools are broken when it comes to the use SSH remote connection, this is a huge problem for me and I guess for some other people as well.
> 
> It would be nice to move on with this. WDYT ?

A darwin contributor is needed to fix on darwin, on linux it works great.

**Comment by drupol** [comment]
Darwin maintainers cannot act as gatekeepers here. I propose to merge, and if there are issues on Darwin, then they should open a PR to fix it.

**Comment by winterqt** [comment]
> Darwin maintainers cannot act as gatekeepers here. I propose to merge, and if there are issues on Darwin, then they should open a PR to fix it.

Who's saying they're... gatekeepers? Seems like a very odd word choice (that has weird implications with it).

If it'a a simple fix, there's no reason it can't be added to this PR. (Note: I have no clue if it is.)

**Comment by uri-canva** [comment]
Re: merging without darwin working, this is why we have the support tiers, see https://github.com/NixOS/rfcs/blob/master/rfcs/0046-platform-support-tiers.md.

Since this problem has been reported to @NixOS/darwin-maintainers in https://github.com/NixOS/nixpkgs/pull/183763#issuecomment-1205487760, I believe there has been enough time and this PR can be merged even while breaking darwin.

I'll remove my request changes to reflect that.

**Comment by winterqt** [comment]
We should probably mark it as broken on Darwin before merging.

**Comment by fabianhjr** [comment]
Marked as broken on darwin

**Comment by drupol** [comment]
> Who's saying they're... gatekeepers? Seems like a very odd word choice (that has weird implications with it).
> 
> If it'a a simple fix, there's no reason it can't be added to this PR. (Note: I have no clue if it is.)
> And if it's not, then it's fine -- just make sure to mark it as broken on Darwin (ideally).

Oops... I apologize for the word I used, I didn't know it could be misinterpreted, I sincerely hope I haven't offended anyone.

What I merely meant here is that if the Darwin team do not reply to solicitations, then we shouldn't wait and move on. I also had the same kind of issue and never got any single reply from the team, so I moved on. This is the reason why I thought that gatekeeper was appropriate, wrongly.

I think marking the package as broken for Darwin is a better idea indeed.

Hope no hard feelings on this and thanks for the quick reaction on this.

**Comment by winterqt** [comment]
> What I merely meant here is that if the Darwin team do not reply to solicitations, then we shouldn't wait and move on. I also had the same kind of issue and never got any single reply from the team, so I moved on. This is the reason why I thought that gatekeeper was appropriate, wrongly.

Agreed -- I didn't realize the call was sent out almost 2 months ago, and that's my apologies.

**Review by khrj — APPROVED** [review]

**Comment by uri-canva** [comment]
Thanks for the patience everyone, sorry for causing some confusion!

**Comment by fabianhjr** [comment]
No worries, hopefully someone with access to the darwin platform can sort the pending issue out.

**Comment by drupol** [comment]
Thanks 🙏

**Pull request merged**
